spec/java/util/OptionalLong.lsl

```diff
@@ -0,0 +1,282 @@
+libsl "1.1.0";
+
+library "std:collections"
+    version "11"
+    language "Java"
+    url "-";
+
+// imports
+
+import "java-common.lsl";
+import "java/lang/_interfaces.lsl";
+import "java/util/function/_interfaces.lsl";
+import "java/util/stream/_interfaces.lsl";
+
+
+// local semantic types
+
+
+
+// automata
+
+@WrapperMeta(
+    src="java.util.OptionalLong",
+    dst="ru.spbpu.libsl.overrides.collections.OptionalLong",
+)
+@public automaton OptionalLong: long
+(
+    var value: long = 0;
+    var present: boolean = false;
+)
+{
+    // states and shifts
+
+    initstate Allocated;
+    state Initialized;
+
+    shift Allocated -> Initialized by [
+        // constructors
+        OptionalLong (),
+        OptionalLong (long),
+
+        // static methods
+        empty,
+        of,
+        ofNullable,
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        getAsLong,
+        isPresent,
+        isEmpty,
+        ifPresent,
+        ifPresentOrElse,
+        filter,
+        map,
+        flatMap,
+        or,
+        stream,
+        orElse,
+        orElseGet,
+        orElseThrow (),
+        orElseThrow (Supplier),
+        toString,
+        hashCode,
+        equals,
+    ];
+
+
+    // constructors
+
+    @private constructor OptionalLong ()
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    @private constructor OptionalLong (x: long)
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    // utilities
+
+    @CacheStaticOnce
+    @static proc _makeEmpty (): OptionalLong
+    {
+        // #problem
+        result = new OptionalLong(state=Initialized);
+    }
+
+
+    @AutoInline
+    proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    // static methods
+
+    @static fun empty (): OptionalLong
+    {
+        result = _makeEmpty();
+    }
+
+
+    @static fun of (x: long): OptionalLong
+    {
+        result = new OptionalLong(state=Initialized, value=x, present=true);
+    }
+
+
+    // methods
+
+    fun equals (other: Object): boolean
+    {
+        if (other == self)
+        {
+            result = true;
+        }
+        else
+        {
+            val isSameType = action OBJECT_SAME_TYPE(self, other);
+            if (isSameType)
+            {
+                val otherValue = OptionalLong(other).value;
+                val otherPresent = OptionalLong(other).present;
+                result = self.value == otherValue && self.present == otherPresent;
```

Comment:
In this class I have found such code:

 return (isPresent && other.isPresent)
               ? Double.compare(value, other.value) == 0
               : isPresent == other.isPresent;
As I understood in this line both fields present can be equals "false":
self.present == otherPresent
Or this is impossible ?

Reply:
Fixed